### DISCUSSION TITLE
On the ability to disable user defined update hooks while preserving realtime API updates

### DISCUSSION THREAD

**Discussion by dschissler** [pocketbase/pocketbase#1993] [Q&A]
I've found a situation where I want to have realtime updates of a record as intensive operations are performed in go routines that are spun off of the update event hook.  I'm able to easily prevent an infinite loop (without using brittle flags) by creating a new Dao.  The problem then is that realtime event hooks aren't fired.

**Comment by ganigeorgiev** [comment]
Unfortunately  I don't think there is an easier way without flags at the moment.

You can try to manually send a message to each `app.SubscriptionsBroker().Clients()` channel but you'll have to perform manually the checks who can receive the message.

I'll consider exporting the internal `apis.broadcastRecord(action, record)` method in the next v0.14/v0.15 release.

**Comment by dschissler** [comment]
I'm using a flag at the moment.  Below is my current strategy.  Perhaps you could bake something into Pocketbase that will be generally useful to everyone?

I think that somehow developers should have more control over what is broadcast for realtime, when PB collection validation available, when model events and custom fancy Record events are available.

For custom Record events something like the pattern that I gave below can be used.  It's nice because the Hook controller doesn't need to know about these interfaces, and you only need to wire it once per collection.

I revised what is passed to these custom Record hooks when I realized that the dao needs to be passed just as the generic Model events.  The reason is for transactions.  In my design the hook controller has the PB App struct embedded.

Importantly!  If you were ever to allow non-DB fields to be received in an event from the API, then it absolutely is not be safe to automatically put it into the Record.  The reason is that these fields could be used as ephemeral flags for things that circumvent important process for niche cases.  So that would be a disaster if the app user could reach deep into the flow like that.

Perhaps a special flag map could be used?  Although that seems like a lot cruft.

```golang
func WireModel(app core.App, tag string, hooks any) {
	if h, ok := hooks.(beforeCreateModelHook); ok {
		app.OnModelBeforeCreate(tag).Add(recordHandler(h.BeforeCreate))
	}
	if h, ok := hooks.(afterCreateModelHook); ok {
		app.OnModelAfterCreate(tag).Add(recordHandler(h.AfterCreate))
	}
	if h, ok := hooks.(beforeUpdateModelHook); ok {
		app.OnModelBeforeUpdate(tag).Add(recordHandler(h.BeforeUpdate))
	}
	if h, ok := hooks.(afterUpdateModelHook); ok {
		app.OnModelAfterUpdate(tag).Add(recordHandler(h.AfterUpdate))
	}
	if h, ok := hooks.(beforeDeleteModelHook); ok {
		app.OnModelBeforeDelete(tag).Add(recordHandler(h.BeforeDelete))
	}
	if h, ok := hooks.(afterDeleteModelHook); ok {
		app.OnModelAfterDelete(tag).Add(recordHandler(h.AfterDelete))
	}
}

type beforeCreateModelHook interface {
	BeforeCreate(e *event.Record) error
}
type afterCreateModelHook interface {
	AfterCreate(e *event.Record) error
}
type beforeUpdateModelHook interface {
	BeforeUpdate(e *event.Record) error
}
type afterUpdateModelHook interface {
	AfterUpdate(e *event.Record) error
}
type beforeDeleteModelHook interface {
	BeforeDelete(e *event.Record) error
}
type afterDeleteModelHook interface {
	AfterDelete(e *event.Record) error
}

func recordHandler(fn hook.Handler[*event.Record]) hook.Handler[*core.ModelEvent] {
	return func(e *core.ModelEvent) error {
		if record, ok := e.Model.(*models.Record); ok {
			if event.AreRecordHooksDisabled(record) {
				return nil
			}
			return fn(&event.Record{
				Dao:    e.Dao,
				Record: record,
			})
		}
		return nil
	}
}
```

```golang
type Hooks struct {
	*sys.Env
}

func (h *Hooks) BeforeCreate(e *event.Record) error {

	return nil
}
```

```golang
const disableCustomRecordHooks = "___disable_custom_record_hooks___"

type Record struct {
	Dao    *daos.Dao
	Record *models.Record
}

func (e *Record) SaveForceClean() error {
	dao := daos.New(e.Dao.DB())
	daoNew := DaoNew(dao)
	return daoNew.SaveRecord(e.Record)
}

func (e *Record) DisableCustomHooks() {
	e.Record.Set(disableCustomRecordHooks, true)
}

func (e *Record) EnableCustomHooks() {
	e.Record.Set(disableCustomRecordHooks, false)
}

func DaoNew(dao *daos.Dao) *daos.Dao {
	return daos.New(dao.DB())
}

func AreRecordCustomHooksDisabled(m *models.Record) bool {
	return m.GetBool(disableCustomRecordHooks)
}
```

**Comment by dschissler** [comment]
@ganigeorgiev Just an idea, what about special flags for model record, and a new form that is only for updates that you allow developers to control what is sent off via realtime events, validation, model event hooks.

Let's say that there was a intensive process taking place in a goroutine that was started in a custom record hook, and it was necessary to send realtime events, but to skip event hooks.  Or the opposite.  Maybe there was going to be a series of updates, but only the final realtime event was desired.  Or what is there weren't any changes happening to the actual record, but instead some meta data was being updated, such a message notifying that some other record had changed.

In my case, I found that there is a particular status in a field that I don't actually care about but I have to set that in order to notify via realtime that something has changed.  Since I already needed to disable record hooks (in any number of ways) to prevent an infinite hook loop then this would be a good time for manual control.

I think that Pocketbase really needs some manner of control like this.  Please think about this sort of thing when you're improving the realtime record binding API.

**Comment by dschissler** [comment]
@ganigeorgiev

Here is my latest pattern for avoiding custom record hooks on updates while keeping collection validation rules and realtime.

Another strategy that I might employ in the future is to call the record hook but to allow it to know if it's going into update hook recursion, or to prevent it.  If I ever need that then I'll go there, otherwise it's just a pointless exercise.

```golang
func NewAdvancedRecordUpsert(app core.App, m *pbmodels.Record) *AdvancedRecordUpsert {
	return &AdvancedRecordUpsert{
		RecordUpsert: forms.NewRecordUpsert(app, m),
		model:        m,
	}
}

type AdvancedRecordUpsert struct {
	*forms.RecordUpsert
	model *pbmodels.Record
}

func (form *AdvancedRecordUpsert) SkipRecordHooks() {
	record_tools.DisableRecordHooks(form.model)
}

func (form *AdvancedRecordUpsert) Submit() error {
	if !record_tools.UsingRecordHooks(form.model) {
		defer record_tools.EnableRecordHooks(form.model)
	}

	return form.RecordUpsert.Submit()
}
```

```golang
func NewHooks(env *sys.Env) *Hooks {
	return &Hooks{Env: env}
}

type Hooks struct {
	*sys.Env
}

func (h *Hooks) BeforeCreate(e *event.Record) error {
	return nil
}

func (h *Hooks) AfterCreate(e *event.Record) error {
	if e.Record.GetString("avatar") == "" {
		return h.generateAvatar(e)
	}

	return nil
}

func (h *Hooks) generateAvatar(e *event.Record) error {
	m := e.Record
	g := govatar.MALE

	b, err := avatar.GenerateForName(g, m.GetString("username"), models.UsersAvatarSize)
	if err != nil {
		return err
	}

	f, err := filesystem.NewFileFromBytes(b, "avatar.jpg")
	if err != nil {
		return errors.New("unable to create avatar file from bytes")
	}

	fups := forms.NewAdvancedRecordUpsert(h.App, m)
	fups.SetDao(e.Dao)
	fups.SkipRecordHooks()
	fups.AddFiles("avatar", f)

	return fups.Submit()
}
```

```golang
const disableRecordHooks = "___disable_record_hooks___bad_boy_@ganigeorgiev_cannot_read_this"

func UsingRecordHooks(m *pbmodels.Record) bool {
	return !m.GetBool(disableRecordHooks)
}

func DisableRecordHooks(m *pbmodels.Record) {
	m.Set(disableRecordHooks, true)
}

func EnableRecordHooks(m *pbmodels.Record) {
	m.Set(disableRecordHooks, false)
}
```

```golang
func modelsInit(env *sys.Env) {
	app := env.App

	hook.WireModel(app, models.Users, user_model.NewHooks(env))
	hook.WireModel(app, models.Articles, article_model.NewHooks(env))
}
```

```golang
func WireModel(app core.App, tag string, hooks any) {
	if h, ok := hooks.(beforeCreateModelHook); ok {
		app.OnModelBeforeCreate(tag).Add(recordHandler(h.BeforeCreate))
	}
	if h, ok := hooks.(afterCreateModelHook); ok {
		app.OnModelAfterCreate(tag).Add(recordHandler(h.AfterCreate))
	}
	if h, ok := hooks.(beforeUpdateModelHook); ok {
		app.OnModelBeforeUpdate(tag).Add(recordHandler(h.BeforeUpdate))
	}
	if h, ok := hooks.(afterUpdateModelHook); ok {
		app.OnModelAfterUpdate(tag).Add(recordHandler(h.AfterUpdate))
	}
	if h, ok := hooks.(beforeDeleteModelHook); ok {
		app.OnModelBeforeDelete(tag).Add(recordHandler(h.BeforeDelete))
	}
	if h, ok := hooks.(afterDeleteModelHook); ok {
		app.OnModelAfterDelete(tag).Add(recordHandler(h.AfterDelete))
	}
}

type beforeCreateModelHook interface {
	BeforeCreate(e *event.Record) error
}
type afterCreateModelHook interface {
	AfterCreate(e *event.Record) error
}
type beforeUpdateModelHook interface {
	BeforeUpdate(e *event.Record) error
}
type afterUpdateModelHook interface {
	AfterUpdate(e *event.Record) error
}
type beforeDeleteModelHook interface {
	BeforeDelete(e *event.Record) error
}
type afterDeleteModelHook interface {
	AfterDelete(e *event.Record) error
}

func recordHandler(fn hook.Handler[*event.Record]) hook.Handler[*core.ModelEvent] {
	return func(e *core.ModelEvent) error {
		if record, ok := e.Model.(*models.Record); ok {
			if record_tools.UsingRecordHooks(record) {
				return fn(&event.Record{
					Dao:    e.Dao,
					Record: record,
				})
			}
		}
		return nil
	}
}
```
